js/src/main/scala/org/soundsofscala/playback/FunctionalAudioPlayer.scala

```diff
@@ -0,0 +1,91 @@
+package org.soundsofscala.playback
+
+import cats.effect.{IO, Ref}
+import cats.syntax.all.*
+import org.scalajs.dom
+import org.scalajs.dom.{AudioBuffer, AudioBufferSourceNode, AudioContext, GainNode}
+import org.soundsofscala.instrument.SampleLoader
+
+case class FunctionalAudioPlayer(
+    path: String,
+    audioBuffer: AudioBuffer,
+    pauseOffset: Ref[IO, Double],
+    sourceNode: Ref[IO, Option[AudioBufferSourceNode]],
+    gainNode: Ref[IO, Option[GainNode]],
+    startTime: Ref[IO, Option[Double]]
+)(using audioContext: AudioContext):
+
+  private val fadeTime = 0.02
+
+  def play(): IO[Unit] =
+    for
+      _ <- stopPlayback()
+      source = audioContext.createBufferSource()
+      gain = audioContext.createGain()
+      offset <- pauseOffset.get
+      _ <- IO {
+        gain.gain.setValueAtTime(0.0, audioContext.currentTime)
+        gain.gain.linearRampToValueAtTime(1.0, audioContext.currentTime + fadeTime)
+        source.buffer = audioBuffer
+        source.connect(gain)
+        gain.connect(audioContext.destination)
+        source.start(0, offset)
+      }
+      _ <- startTime.set((audioContext.currentTime - offset).some)
+      _ <- sourceNode.set(source.some)
+      _ <- gainNode.set(gain.some)
+    yield ()
+
+  def stop(): IO[Unit] =
+    for
+      _ <- stopPlayback()
+      _ <- pauseOffset.set(0.0)
+      _ <- startTime.set(None)
+    yield ()
```

Comment:
yeah, this is the reason why I'm thinking that perhaps using Resource might be the right choice here? I might be wrong though, as I'm not sure about the implications with the JS runtime